artemis-server/src/test/resources/ConfigurationTest-full-config.xml

```diff
@@ -378,7 +378,11 @@
 
       <address-settings>
          <address-setting match="#">
-            <dead-letter-address-prefix>DLA.</dead-letter-address-prefix>
+            <dead-letter-address-auto-create prefix="DLA.">
+               <routing-type>AS_ORIGIN</routing-type>
```

Comment:
Ditto about other comments on using null to represent use the parent.

Reply:
I suppose AS_ORIGIN might be a misleading name.
What it does is that it makes sure messages are delivered to corresponding queues.
Imagine you have one address with 10 queues with multicast and consumption from q10 fails. So what it does it will deliver that failed message to DLA.q10 only, but not any others in DLA.
Same for anycast as this will ignore anycast routing and deliver to DLA.q10.
So it is not about taking routing from parent destination but rather providing predictive DL message delivery to corresponding DLQ instead of routed with ANYCAST or MULTICAST.

Reply:
So an address can be of type multicast or anycast. By default if the queue that dlq is occuring on is a multicast, it should send to address as multicast. And vice versa


Re this pr youre are going to have one dla per queue right. So simply be default the malling shoild be if queue is multi the dla should be multi, if queue is anycast, then dla should be anycast.

Reply:
To achieve this simply this becomes an override setting, and if not set the default behaviour described occurs. Thus meaning you only need two state enum for the override and the third state which do not override and is default matching the routing type of the queue, by not setting/being null

Reply:
I have a feeling there misunderstanding what AS_ORIGIN does due to very misleading name.
In fact, I have renamed AS_ORIGIN in my local to CORRESPONDING_QUEUE.
So the 3rd state is unlike all others and it is not behaving like ANYCAST nor MULTICAST.
It works in a completely different way and it is a genuine 3rd type of routing- an exact match which can be considered as MULTICAST with filters.
Also it can be considered as strict routing, where unlike ANYCAST or MULTICAST the message is delivered always to the exact corresponding DLQ, in fact, regardless of ANYCAST or MULTICAST and unlike them.
So let's assume we have a scenario where we have address A with queues q1, q2, q,3.
Let's assume q2 has 10 messages, while q1 and q3 have 1 and that all messages fail to consume, so this is 12 messages total.
In ANYCAST scenario, DLA.q1, DLA.q2, DLA.q3 each will be delivered 4 messages- still 12 total.
In MULTICAST scenario, each will be delivered 12 messages, so 36 total.
In AS_ORIGIN or new name CORRESPONDING_QUEUE, DLA.q1 will receive 1, DLA.q2 will receive 10 and DLA.q3 will receive 1.
I hope above example represents the difference between the 3rd state and 2 known states.

Reply:
So now youve lost me and kinda -1 it. I was under the impression:

Dlqprefix=dlq

Given 
Address: foo
Queues: bar and bar2

There then would be 
deadletter address bob.foo

With queues

Dlq.bar and dlq.bar2

Where the dlqs would be of same type as the parent.

And if a message in bar it would be dead letter routed to goto dlq.bar under dlq.foo

Like wise a message in bar2 would dlq into dlq.bar2 under address bob

So essentially we have a one to one mapping of address and queues with a dead letter equiv. 

Using the ability to route directly to specific queues, via fqqn or queue id.

This was my original understanding and what i was +1 for

Reply:
Remember internal architecture and address and queue model is very different to activemq classic.

Reply:
@michaelpearce-gain looks like we have now proper discussion :)
I am glad we have reached this point and I hope it is not a problem for you.

Regarding above:

> Dlqprefix=dlq
> 
> Given
> Address: foo
> Queues: bar and bar2
> 
> There then would be
> deadletter address dlq.foo
> 
> With queues
> 
> Dlq.bar and dlq.bar2 

Correct

> (where these would be of same routing type of the parent queue by default)
> Where the dlqs would be of same type as the parent.

Depends on routing config, unless null then yes (in my local code)

> And if a message in bar queue it would be dead letter routed to goto dlq.bar under address dlq.foo (eg we can route this by fqqn of dlq.foo::dlqbar)

Yes and no. In anycast case if DLA would have more than one queue, it would land in one of them. In multicast case it would land in all of them. And that is why there is 3rd state, which does routing by fqqn directly to dlq.foo::dlq.bar.

And I think this is the point where we are having misunderstanding.
My understanding was that as default we want to deliver the dead message to non fqqn address, where routing can be applied.
With option of using FQQN as a 3rd state.
I have a feeling, that your expectation is that messages should be always delivered using FQQN, regardless of routing type.

Is that what you mean?

Also should we move the conversation to some sort of IRC?

Reply:
3rd state removed and now using FQQN address always for DL message delivery to DLQ.
This code is not pushed yet.
I will create new PR today to master as requested.

Thanks for your comments and clarifications.

Reply:
Still -1 sorry you missed the point here. The routing type is the type that the dlq queue has when created. Yes always fqqn when using dynamic dla/dlq. 



I use slack not irc.

Reply:
Having this one to one mapping will make other bits easier, like what occurs when parent queue is deleted, simply the dlq queue is deleted. Etc. You can map lifecycles alot easier, like wise queue settings (just simply you allow overrides on those)

Reply:
@michaelpearce-gain 
I am not familiar with slack. How can I find you there as I think it would be very beneficial to be able to have real time conversation rather than comments.

Reply:
> Yes always fqqn when using dynamic dla/dlq.
> If you override it, youre simply changing thr routing type of the queue that is created. You still dqqn to the dynamic.

That clarifies the misunderstanding, thanks.

> when parent queue is deleted, simply the dlq queue is deleted.

Assuming this is your requirement for this functionality? Or just an example?

Reply:
So these all the issues that can crop up. 

Remember when adding features we need to think how could it be disastrous for someone, as a community we need to support it. 

Thus all the edge cases ive already on the PR to apache (currently cleberts) commented on MUST be addressed.

E.g. think about it you wouldnt want dlqs hanging about, you could end up with a disaster amount of them as clients create and destroy them over time, this occurs alot where auto generated queues with jms consumers on topics (temp queue). 

You need to think about all these edge scenarios requirement of yours or not.

Reply:
I think above is now clear.
I have implemented that already in my local.
Also added this as requirements in https://jira.apache.org/jira/browse/ARTEMIS-2420.

Resolving this conversation as I need to close this PR and will be creating new one.